src/client/dfuse/ops/open.c

```diff
@@ -203,16 +304,22 @@ dfuse_cb_release(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi)
 		}
 	}
 	DFUSE_TRA_DEBUG(oh, "il_calls %d, caching %d,", il_calls, oh->doh_caching);
-	if (il_calls != 0) {
+	if (il_calls != 0)
 		atomic_fetch_sub_relaxed(&oh->doh_ie->ie_il_count, 1);
-	}
-	atomic_fetch_sub_relaxed(&oh->doh_ie->ie_open_count, 1);
+	oc = atomic_fetch_sub_relaxed(&oh->doh_ie->ie_open_count, 1);
 
 	if (oh->doh_evict_on_close) {
 		ie = oh->doh_ie;
 		atomic_fetch_add_relaxed(&ie->ie_ref, 1);
 	}
 
+	if (update_data_cache) {
+		dfuse_dcache_set_time(oh->doh_ie);
+		if (oc == 1)
+			/* Fire off a ostatx call to fetch the new mtime */
+			size_resample(dfuse_info, oh->doh_ie);
```

Comment:
a stat on a file close is going to be pretty bad for workloads that have millions of files ..

Reply:
This is only for files which were written to.  If there's any access to the file afterwards then the kernel would do this on next access anyway so the only additional network load here is for files which are written to but then not accessed again.

Reply:
most of workloads we have seen today write/read to the file with 1 handle. i.e. there is no close and re-access.
What workload are you thinking about here that does this write, close, open, read that would benefit from this?

Reply:
Most applications do not read the files in a short time after writing files. Some applications (e.g., computation chemistry) may do write & close then open & read later because intermediate data are too large to fit in memory.

Reply:
You're right this will help with some I/O patterns and not others, I don't think we can speculate accurately on which is more common but the bug is clear - with current master the cache is dropped on open in the open/write/close -> open/read/close case which is a simple bug due to having the kernel handle cache validity decisions based on incorrect assumptions on how mtime updates are handled.

Some applications and workflows won't be triggering this bug and will therefore be unaffected by it, a lot more will be triggering it and benefit from the fix, interactive workloads will probably benefit the most but plenty of HPC ones will.  IOR certainly would if using the `-C`/reorder flag and the right rank allocation (although this option is mainly there to force clients onto other nodes to avoid the use of in-node caching).

Reply:
my main concern is the effect of an extra stat on those unaffected workloads. i do not agree that they are "unaffected".
anyway since you mentioned IOR is a use case that can benefit, let's gather some results then before we decide on this fix.

Reply:
I did have some numbers on this when I was developing it and it was showing a very high proportion of these speculative stat calls were returned to the kernel - but it'll be entirely application dependant as you say.

Anyway - the extra stat is asynchronous but launched on close so it'll be impossible to measure from a client perspective as the extra cost, if it is indeed extra, is all server side so won't show up in any simple benchmark.